fdbserver/DataDistribution.actor.cpp

```diff
@@ -4061,103 +4136,88 @@ ACTOR Future<std::vector<std::pair<StorageServerInterface, ProcessClass>>> getSe
 
-						if (nextIt->second.size() &&
-						    nextIt->second[0]->lastKnownInterface.locality.get(localityKey) == localityValue) {
-							writeValue = nextIt->first; // locality matched
-							break;
-						}
-						serverInfoSize--;
-						if (!serverInfoSize) {
-							// None of the entries in pid2server_info matched the given locality.
-							writeValue = LiteralStringRef("");
-							TraceEvent("PerpetualNextWigglingStoragePIDNotFound", teamCollection->distributorId)
-							    .detail("WriteValue", "No process matched the given perpetualStorageWiggleLocality")
-							    .detail("PerpetualStorageWiggleLocality", localityKeyValue);
-							break;
-						}
-						nextIt++;
-					}
-				}
+			// TraceEvent("PerpetualLocality").detail("Server", server->lastKnownInterface.locality.get(localityKey)).detail("Desire", localityValue);
+			if (server->lastKnownInterface.locality.get(localityKey.get()) == localityValue) {
+				return id.get();
 			} else {
-				Optional<Value> value = wait(tr.get(writeKey));
-				Value pid = teamCollection->pid2server_info.begin()->first;
-				if (value.present()) {
-					auto nextIt = teamCollection->pid2server_info.upper_bound(value.get());
-					if (nextIt == teamCollection->pid2server_info.end()) {
-						writeValue = pid;
-					} else {
-						writeValue = nextIt->first;
-					}
-				} else {
-					writeValue = pid;
+				if (teamCollection->storageWiggler->empty()) {
+					// None of the entries in wiggle queue matches the given locality.
+					TraceEvent("PerpetualNextWigglingStorageIDNotFound", teamCollection->distributorId)
+					    .detail("WriteValue", "No process matched the given perpetualStorageWiggleLocality")
+					    .detail("PerpetualStorageWiggleLocality",
+					            teamCollection->configuration.perpetualStorageWiggleLocality);
 				}
+				continue;
 			}
+		} else {
+			auto id = teamCollection->storageWiggler->getNextServerId();
+			if (!id.present())
+				continue;
+			return id.get();
+		}
+	}
+}
 
-			tr.set(writeKey, writeValue);
+// Create a transaction updating `wigglingStorageServerPrefix` to the next serverID according to a sorted wiggle_pq
+// maintained by the wiggler.
+ACTOR Future<Void> updateNextWigglingStorageID(DDTeamCollection* teamCollection) {
+	state ReadYourWritesTransaction tr(teamCollection->cx);
+	state const Key writeKey =
+	    wigglingStorageServerPrefix.withSuffix(teamCollection->primary ? "primary"_sr : "remote"_sr);
+	state UID nextId = wait(getNextWigglingServerID(teamCollection));
+
+	loop {
+		// write the next server id
+		try {
+			tr.setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+			tr.set(writeKey, BinaryWriter::toValue(nextId, Unversioned()));
 			wait(tr.commit());
 			break;
 		} catch (Error& e) {
 			wait(tr.onError(e));
 		}
 	}
-	TraceEvent(SevDebug, "PerpetualNextWigglingStoragePID", teamCollection->distributorId)
+	TraceEvent(SevDebug, "PerpetualNextWigglingStorageID", teamCollection->distributorId)
```

Comment:
nit: "PerpetualNextWigglingStorageID" -> "PerpetualWiggleNextStorageID"..

A general comment: "wiggle" "wiggling" and other variants are used interchangeably; I find a lot of value in recording logs having common prefix i.e. "PerpetualStorageWiggle_XX" Or "PerpetualStorageWiggleXX"; if needed just searching "PerpetualStorageWiggle" would given all related logs. Thoughts?